eth_account/messages.py

```diff
@@ -432,10 +443,107 @@ def encode_typed_data(
+        >>> full_message = {
+        ...     "types": {
+        ...         "EIP712Domain": [
+        ...             {"name": "name", "type": "string"},
+        ...             {"name": "version", "type": "string"},
+        ...             {"name": "chainId", "type": "uint256"},
+        ...             {"name": "verifyingContract", "type": "address"},
+        ...             {"name": "salt", "type": "bytes32"},
+        ...         ],
+        ...         "Person": [
+        ...             {"name": "name", "type": "string"},
+        ...             {"name": "wallet", "type": "address"},
+        ...         ],
+        ...         "Mail": [
+        ...             {"name": "from", "type": "Person"},
+        ...             {"name": "to", "type": "Person"},
+        ...             {"name": "contents", "type": "string"},
+        ...         ],
+        ...     },
+        ...     "primaryType": "Mail",
+        ...     "domain": {
+        ...         "name": "Ether Mail",
+        ...         "version": "1",
+        ...         "chainId": 1,
+        ...         "verifyingContract": "0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC",
+        ...         "salt": b"decafbeef"
+        ...     },
+        ...     "message": {
+        ...         "from": {
+        ...             "name": "Cow",
+        ...             "wallet": "0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826"
+        ...         },
+        ...         "to": {
+        ...             "name": "Bob",
+        ...             "wallet": "0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB"
+        ...         },
+        ...         "contents": "Hello, Bob!",
+        ...     },
+        ... }
+        >>> signable_msg_2 = encode_typed_data(full_message=full_message)
+        >>> signed_msg_2 = Account.sign_message(signable_msg, key)
+        >>> signed_msg_2.messageHash
+        HexBytes('0xc5bb16ccc59ae9a3ad1cb8343d4e3351f057c994a97656e1aff8c134e56f7530')
+        >>> signed_msg_2 == signed_msg
+        True
     .. _EIP-712: https://eips.ethereum.org/EIPS/eip-712
     """
+    if full_message is not None and (
+        domain_data is not None or message_types is not None or message_data is not None
+    ):
+        raise ValueError(
+            "You may supply either `full_message` or `domain_data`, `message_types`, "
+            "and `message_data`, but not both."
+        )
+
+    if full_message is not None:
+        # If EIP712Domain types were provided, check that they match the domain data
+        if "EIP712Domain" in full_message["types"]:
+            domain_data_keys = full_message["domain"].keys()
+            domain_types_keys = [
+                field["name"] for field in full_message["types"]["EIP712Domain"]
+            ]
+
+            if set(domain_data_keys) != (set(domain_types_keys)):
+                raise ValidationError(
+                    "The fields provided in ``domain`` do not match the fields provided"
```

Comment:
I haven't seen a lot of messages raised with any readable formatting, it's usually just a long string. Curious if some newlines in the exceptions would make it easier to line up the two sets for better readability.

Reply:
This is going to refer only to the keys within the domain section of the message, of which there are only 5 valid options. The full error message will be something like: 

```py
ValidationError: The fields provided in `domain` do not match the fields provided in `types.EIP712Domain`. The fields provided in `domain` were `['name', 'version', 'chainId', 'verifyingContract', 'salt']`, but the fields provided in `types.EIP712Domain` were `['name', 'version', 'chainId', 'verifyingContract']`.
```

I do like the idea of newlines to aid comparison, but it seems overkill for comparing lists of 5 or fewer items. Leaving as-is for now, but happy to discuss further!